File: data_structures/02list/merge_list/merge_two_sorted_list.cpp
/*
Merge two sorted linked lists and return it as a new list. 
The new list should be made by splicing together the nodes 
of the first two lists.

合并两个非降序链表的并集。
1->2->3 和 2->3->5 为
1->2->3->5. 只能输出结果，
不能修改两个链表的数据

Time Complexity O(m+n)
*/



/* CPP */

#include <iostream>
#include <string>
using namespace std;


// Definition for singly-linked list.

struct ListNode {
	int val;
	ListNode *next;
	ListNode(int x) : val(x), next(NULL) {}
};

class Solution {

public:
    ListNode* mergeTwoLists(ListNode  *l1, listNode *l2) {
       if (l1 == NULL) {
            return l2;
        }

        if (l2 == NULL) {
            return l1;
        }

        ListNode *mergedListHead = NULL;

        if (l1->val < l2->val) {
            mergedListHead = l1;
            mergedListHead->next = mergeTwoLists(l1->next, l2);
        } else if (l1->val >= l2->val) {
            mergedListHead = l2;
            mergedListHead->next = mergeTwoLists(l1, l2->next);
        }

        return mergedListHead;        
    }
};

File: 02list/merge_lists/MergeTwoSortedList.java
/*
Merge two sorted linked lists and return it as a new list. 
The new list should be made by splicing together the nodes 
of the first two lists.


合并两个非降序链表的并集。
1->2->3 和 2->3->5 为
1->2->3->5. 只能输出结果，
不能修改两个链表的数据

Time Complexity O(m+n)
*/


/* Java */

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */

public class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if (l1 == null) {
            return l2;
        }
        
        if (l2 == null) {
            return l1;
        }
        
        ListNode head = null;
        
        if (l1.val < l2.val) {
            head = l1;
            head.next = mergeTwoLists(l1.next, l2);
        } else {
            head = l2;
            head.next = mergeTwoLists(l1, l2.next);
        }
        
        return head;
    }
}